### ISSUE TITLE
[Domino Visual Editor] Interface Similar to  "Objects" from Domino Designer

### ISSUE THREAD

**Issue by JoelProminic** [Moonshine-IDE/Moonshine-IDE#1009]
@JustinProminic requested that we add a feature similar to the Objects interface from Domino Designer.   This interface is normally found in an expandable area below the form editor:

![image](https://user-images.githubusercontent.com/24532731/160937852-01734578-8239-49ce-a385-23909847051a.png)

This interface holds some relevant properties for Forms (and other object types, as we add support for them).

This is a low priority task for now, but I have added it for tracking.

**Comment by JoelProminic** [comment]
Here are some GUI examples.

Here is the full Object list for Global and Forms.  We will revisit the other categories (Fields, Subforms, etc) as a later step:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/450ba7aa-7361-40c7-b7d3-e4e0dd0233ca)

The icon indicates the type of the field:
- Scroll - Lotuscript (i.e. `(Options)`)
- Diamond:  Formula (i.e. `WebQueryOpen`)
- Circle:  Javascript (i.e. `JS Header`)

The icons are filled in if the Object has been populated.

**Comment by JoelProminic** [comment]
I set values for the different Global and Form-level Objects.  Here is the exported DXL:
[2023_11_03__TestFormDefaultLanguages.dxl.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/13254459/2023_11_03__TestFormDefaultLanguages.dxl.zip)

Here are the default values I found while setting up the test form:

Object | Default Value
-- | --
Globals | \-
(Options) | `Option Public`
(Initialize) | `Sub Initialize`
(Terminate) | `Sub Terminate`
Form | \-
WebQueryOpen | `@Command([ToolsRunMacro]; "<Your agent goes here>")`
WebQuerySave | `@Command([ToolsRunMacro]; "<Your agent goes here>")`
QueryOpen | `Sub Queryopen(Source As Notesuidocument, Mode As Integer, Isnewdoc As Variant, Continue As Variant)`
PostOpen | `Sub Postopen(Source As Notesuidocument)`
Querymodechange | `Sub Querymodechange(Source As Notesuidocument, Continue As Variant)`
Postmodechange | `Sub Postmodechange(Source As Notesuidocument)`
Queryrecalc | `Sub Queryrecalc(Source As Notesuidocument, Continue As Variant)`
Postrecalc | `Sub Postrecalc(Source As Notesuidocument)`
Querysave | `Sub Querysave(Source As Notesuidocument, Continue As Variant)`
Postsave | `Sub Postsave(Source As Notesuidocument)`
Querysend | `Sub Querysend(Source As Notesuidocument, Continue As Variant)`
Postsend | `Sub Postsend(Source As Notesuidocument)`
Queryclose | `Sub Queryclose(Source As Notesuidocument, Continue As Variant)`
Initialize | `Sub Initialize`
Terminate | `Sub Terminate`

Note that all `Sub` defaults end with an blank line and `End Sub`

```
Sub Queryopen(Source As Notesuidocument, Mode As Integer, Isnewdoc As Variant, Continue As Variant)
	
End Sub
```

TODO:  Add examples for properties which support different languages

**Comment by JoelProminic** [comment]
I added a second form to show the Objects with alternative environments or languages.  For the languages, I used LotusScript when available to show the default value (exclude the comment). 

[2023_11_03__TestFormOtherLanguages1.dxl.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/13254663/2023_11_03__TestFormOtherLanguages1.dxl.zip)

Object | Environments | Languages
-- | -- | --
JS Header | Web/Client | Javascript/Common Javascript
onHelp | Web/Client | JavaScript/Common JavaScript
onHelp | Web/Client | JavaScript/Common JavaScript
onLoad | Web/Client | Formula/LotusScript/JavaScript/Common JavaScript
onUnload | Web/Client | Formula/LotusScript/JavaScript/Common JavaScript
Onsubmit | Web/Client | Formula/LotusScript/JavaScript/Common JavaScript
Queryopen | Client | Formula/Lotuscript/JavaScript
Postopen | Client | Formula/Lotuscript
Querymodechange | Client | Formula/Lotuscript/JavaScript
Postrecalc | Client | Formula/Lotuscript/JavaScript
Querysave | Client | Formula/Lotuscript
Postsave | Client | Formula/Lotuscript/JavaScript
Querysend | Client | Formula/Lotuscript/JavaScript
Queryclose | Client | Formula/Lotuscript

Note that the icon changes based on the language.

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/a1833613-75b1-43d1-ae90-93af4984e4dd)

**Comment by JoelProminic** [comment]
Some issues I found from my tests today:
- "HTML Head Conten" > "HTML Head Content"
- "(Globals)Email Reply Common Sub" is hard-coded.  Should be "(Globals)%Form%" (no space after "(Globals")
- "Email Reply Common Sub(Sub)" is hard-coded.  Should be "%Form% (Form) " (there is space before "(Form)")
- Code section should always use fixed-width/monospace/Courier font.  We should probably reuse the code editor for this (see below)

Some suggests from @rat-moonshine for using the Code editor:

>10:08:12: What type of component it currently has? 
>10:08:15: TextArea? 
>10:08:51: Okay, of course he can use the FeathersUI.TextEditor as a simple text-editor - same way we're using in Native or GBAuth 
>10:09:08: Let me find some quick code for him to use. 
>10:16:08: Here's a pseudo code that should work:
>
>    this.myTextEditor = new TextEditor(null, true);
>    this.myTextEditor.wordWrap = true;
>    this.myTextEditor.showLineNumbers = false;
>    var wrapper:FeathersUIWrapper = new FeathersUIWrapper(this.myTextEditor);
>    wrapper.percentWidth = wrapper.percentHeight = 100;
>    wrapper.minWidth = wrapper.minHeight = 0;
>    textContainer.addElementAt(wrapper, 0);
>
>    myTextEditor.text = $text;

Try setting `wordWrap=false` and `showLineNumbers=true` to make this more like the code editor.  We can change these later if they don't look good in this context

**Comment by rat-moonshine** [comment]
I see I supplied the code as TextEditor read-only (non-editable). Make adjustment as you may need)

`new TextEditor(?text:String=null, ?readOnly:Bool=false)`

**Comment by feather812002** [comment]
@rat-moonshine Thanks ,I will try it .

**Comment by feather812002** [comment]
@JoelProminic  are you sure the second  is "%Form% (Form) "  rather than  "%Form% (Sub)"?

**Comment by JoelProminic** [comment]
@feather812002, you can see an example for "(Form)" [here](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1009#issuecomment-1793055068).   The screenshot from the original description  was for a subform, so the suffix was "(Subform)" (this was cut off).

**Comment by feather812002** [comment]
Form script : 

```
<item name='$$ScriptName' summary='false' sign='true'><text>TestFormDefaultLanguages</text></item>
<item name='$$FormScript' summary='false' sign='true'><text>'++LotusScript Development Environment:2:5:(Options):0:74
' test Options
'++LotusScript Development Environment:2:5:(Forward):0:1
Declare Sub Queryopen(Source As Notesuidocument, Mode As Integer, Isnewdoc As Variant, Continue As Variant)
Declare Sub Postopen(Source As Notesuidocument)
Declare Sub Querymodechange(Source As Notesuidocument, Continue As Variant)
Declare Sub Postmodechange(Source As Notesuidocument)
Declare Sub Queryrecalc(Source As Notesuidocument, Continue As Variant)
Declare Sub Postrecalc(Source As Notesuidocument)
Declare Sub Querysave(Source As Notesuidocument, Continue As Variant)
Declare Sub Postsave(Source As Notesuidocument)
Declare Sub Querysend(Source As Notesuidocument, Continue As Variant)
Declare Sub Queryclose(Source As Notesuidocument, Continue As Variant)
Declare Sub Postsend(Source As Notesuidocument)
Declare Sub Initialize
Declare Sub Terminate

'++LotusScript Development Environment:2:5:(Declarations):0:10
' Test Declarations
'++LotusScript Development Environment:2:2:BindEvents:1:129
Private Sub BindEvents(Byval Objectname_ As String)
	Static Source As NOTESUIDOCUMENT
	Set Source = Bind(Objectname_)
	On Event Queryopen From Source Call Queryopen
	On Event Postopen From Source Call Postopen
	On Event Querymodechange From Source Call Querymodechange
	On Event Postmodechange From Source Call Postmodechange
	On Event Queryrecalc From Source Call Queryrecalc
	On Event Postrecalc From Source Call Postrecalc
	On Event Querysave From Source Call Querysave
	On Event Postsave From Source Call Postsave
	On Event Querysend From Source Call Querysend
	On Event Queryclose From Source Call Queryclose
	On Event Postsend From Source Call Postsend
End Sub

'++LotusScript Development Environment:2:2:Queryopen:1:12
Sub Queryopen(Source As Notesuidocument, Mode As Integer, Isnewdoc As Variant, Continue As Variant)
	' Test QueryOpen
End Sub
'++LotusScript Development Environment:2:2:Postopen:1:12
Sub Postopen(Source As Notesuidocument)
	' Test PostOpen
End Sub
'++LotusScript Development Environment:2:2:Querymodechange:1:12
Sub Querymodechange(Source As Notesuidocument, Continue As Variant)
	' Test Querymodechange
End Sub
'++LotusScript Development Environment:2:2:Postmodechange:1:12
Sub Postmodechange(Source As Notesuidocument)
	' Test Postmodechange
End Sub
'++LotusScript Development Environment:2:2:Queryrecalc:1:12
Sub Queryrecalc(Source As Notesuidocument, Continue As Variant)
	' test Queryrecalc
End Sub
'++LotusScript Development Environment:2:2:Postrecalc:1:12
Sub Postrecalc(Source As Notesuidocument)
	' test Postrecalc
End Sub
'++LotusScript Development Environment:2:2:Querysave:1:12
Sub Querysave(Source As Notesuidocument, Continue As Variant)
	' Test Querysave
End Sub
'++LotusScript Development Environment:2:2:Postsave:1:12
Sub Postsave(Source As Notesuidocument)
	' Test Postsave
End Sub
'++LotusScript Development Environment:2:2:Querysend:1:12
Sub Querysend(Source As Notesuidocument, Continue As Variant)
	' Test Querysend
End Sub
'++LotusScript Development Environment:2:2:Queryclose:1:12
Sub Queryclose(Source As Notesuidocument, Continue As Variant)
	' Test Queryclose
End Sub
'++LotusScript Development Environment:2:2:Postsend:1:12
Sub Postsend(Source As Notesuidocument)
	' Test Postsend
End Sub
'++LotusScript Development Environment:2:2:Initialize:1:10
Sub Initialize
	' test initialize
End Sub
'++LotusScript Development Environment:2:2:Terminate:1:10
Sub Terminate
	' Test terminate
End Sub<break/></text></item>
```

From script compile


```
<item name='$$$FormScript_O' sign='true'>
<rawitemdata type='14'>
AQAAAkxTT0IDABQIZW4AAAQAMQDaAAQAAAA4CwgDJAm8CQQAwADMAAAAAAAkAGwBGAA4ACwBiAAA
AAAAAAAAAA4AAAAYALwJAAAAAAAAAAAAAAAARAJEAqQDpAMEBcQHTARMBGQGJAkAAAAAAAAAAGwB
vAnsAgwHGAB8CAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAEAAADMAMwAzADMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAEABQBMb3R1cyBQcm9kdWN0AAAAVjAuMAAAAAAAAAAAAAAAAAAA
AAAEAAAA2gAAAAAAAAA4CwAACAMAAAAAAAAAAAAAAQAAAAAAAAABAAAAAAAAAAEAAAAAAAAAAAAA
ABEAAABXMzJJJQMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAwAMAwAAAABUAAkAKgAxAEYAMwAwADIAOAAxADgAAABsAAMATgBFAFcAAACAAAYARABFAEwA
RQBUAEUAAAAAAFQBCgBJAE4ASQBUAEkAQQBMAEkAWgBFAAAAAABUAgkAVABFAFIATQBJAE4AQQBU
AEUAAADIAQYATwBCAEoARQBDAFQAAAAAANAAAAAAAAAAiAEhACgARwBMAE8AQgBBAEwAUwApAFQA
RQBTAFQARgBPAFIATQBEAEUARgBBAFUATABUAEwAQQBOAEcAVQBBAEcARQBTAAAAGAEhACgARwBs
AG8AYgBhAGwAcwApAFQAZQBzAHQARgBvAHIAbQBEAGUAZgBhAHUAbAB0AEwAYQBuAGcAdQBhAGcA
ZQBzAAAAPAEGAFMATwBVAFIAQwBFAAAAAAD//wQATQBPAEQARQAAAAAA3AEIAEkAUwBOAEUAVwBE
AE8AQwAAAAAApAEIAEMATwBOAFQASQBOAFUARQAAAAAA1AIKAEIASQBOAEQARQBWAEUATgBUAFMA
AAAAADACCwBPAEIASgBFAEMAVABOAEEATQBFAF8AAAD0AQ8ATgBPAFQARQBTAFUASQBEAE8AQwBV
AE0ARQBOAFQAAAAMAgYAJQBMAFMAWABVAEkAAAAAALwCCQBRAFUARQBSAFkATwBQAEUATgAAAP//
CABQAE8AUwBUAE8AUABFAE4AAAAAAHACDwBRAFUARQBSAFkATQBPAEQARQBDAEgAQQBOAEcARQAA
AIwCDgBQAE8AUwBUAE0ATwBEAEUAQwBIAEEATgBHAEUAAAAAAP//CwBRAFUARQBSAFkAUgBFAEMA
QQBMAEMAAAD//woAUABPAFMAVABSAEUAQwBBAEwAQwAAAAAApAIJAFEAVQBFAFIAWQBTAEEAVgBF
AAAA//8IAFAATwBTAFQAUwBBAFYARQAAAAAA//8JAFEAVQBFAFIAWQBTAEUATgBEAAAA8AIKAFEA
VQBFAFIAWQBDAEwATwBTAEUAAAAAAP//CABQAE8AUwBUAFMARQBOAEQAAAAAAAUAPAsAAAAAZuU6
ExgAAAAAAIwAQNr9HwAAAAAAAAAACAAAAGwBfAhwAQAAAAAAAAIAAACwAFwBAAAAAAAAAAAAAAAA
AAAAALAAsAAAAAAAAAAAAAAAAAAAAAAAAAAAAFwBXAEAAAAAAAAAAAEAAAACAAAABgAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAADAAEAXAEAAIwBBggAAAAAGgAAAMwB//8AAAAAEAAAAAAAAACoAQAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAQVEyntLmkQ
v10A3QERhrcAAAAAAAAAAAAAAAAAAAAAZAAAAOBI4m0AAAAAAAAAAAIAwAAAACgLAAAAAAoAAQD/
/wAAAwAgAAAAAAAcAQkCAADMAAgAAABEArwJ4AEAAAAAAAAEAAAABAI0AgAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAA0AjQCJAIkAgAAAAAEAgQCFAIUAgAAAAAEAAAAAAAAAKYAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AwABABQCAAAcAQkCAADMAAMAAQAkAgAAMAEBAAEAAAADAAEANAIAAEABAAACAAAAAwABAAAAAABY
AQAAAwAAAAgAAADsAgAA+AEAAAAAAAABAAAA3ALcAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAADcAtwCAAAAAAAAAAABAAAAAAAAAKoAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAwABAAAAAAAcAQkC
AADMAAgAAACkAwwHEAIAAAAAAAACAAAAhAOUAwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACU
A5QDAAAAAAAAAACEA4QDAAAAAAAAAAACAAAAAAAAAK4AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAwABAJQDAAAcAQkCAADM
AAMAAQAAAAAAWAEAAAEAAAAIAAAATAQAADQCAAAAAAAAAQAAADwEPAQAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAPAQ8BAAAAAAAAAAAAQAAAAAAAACyAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMA
AQAAAAAAHAEJAgAAzAAIAAAABAUAAFgCAAAAAAAAAgAAAOQE9AQAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAA9AT0BAAAAAAAAAAA5ATkBAAAAAAAAAAAAgAAAAAAAAC2AAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMAAQD0
BAAAHAEJAgAAzAADAAEAAAAAAFgBAAABAAAACAAAAKwFrAV0AgAAAAAAAAEAAACcBZwFAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAJwFnAUAAAAAAAAAAAEAAAAAAAAAugAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAADAAEAAAAAABwBCQIAAMwACAAAAGQGxAeQAgAAAAAAAAIAAABEBlQGAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAFQGVAYAAAAAAAAAAEQGRAYAAAAAAAAAAAIAAAAAAAAAvgAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAADAAEAVAYAABwBCQIAAMwAAwABAAAAAABYAQAAAQAAAAgAAAAMByQJqAIAAAAAAAABAAAA
/Ab8BgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAD8BvwGAAAAAAAAAAAB
AAAAAAAAAMIAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAwABAAAAAAAcAQkCAADMAAgAAADEBwAAwAIAAAAAAAACAAAApAe0
BwAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAC0B7QHAAAAAAAAAACkB6QHAAAAAAAAAAACAAAA
AAAAAMYAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAwABALQHAAAcAQkCAADMAAMAAQAAAAAAWAEAAAEAAAAIAAAAfAgAANgC
AAAAAAAAAgAAAFwIbAgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAbAhsCAAAAAAAAAAAXAhc
CAAAAAAAAAAAAgAAAAAAAADKAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMAAQBsCAAAHAEJAgAAzAADAAEAAAAAAFgBAAAB
AAAACAAAACQJAAD0AgAAAAAAAAEAAAAUCRQJAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAABQJFAkAAAAAAAAAAAEAAAAAAAAAzgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADAAEAAAAAABwBCQIAAMwA
CAAAALwJAAA8AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA0gAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAIAAAAAAAAAFgAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAADWAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAABMAAADgAQAAygAAAAMAAAADAAAAAAABCAAIAAATAFQK+AEAAMsA
AAAAAAAAEwBwChACAADNAAAAAQAAAAEAAAAAAAAAEwCACjQCAADOAAAAAAAAABMAmApYAgAA8AAA
AAEAAAABAAAAAAAAABMAqAp0AgAAzwAAAAAAAAATAMAKkAIAANAAAAABAAAAAQAAAAAAAAATANAK
qAIAANEAAAAAAAAAEwDoCsACAADuAAAAAQAAAAEAAAAAAAAAEwD4CtgCAADMAAAAAQAAAAEAAAAA
AAAAEwAQC/QCAADvAAAAAAAAAAQA3gAAAAAA0tQAHQAAGhcAW1wBSbAAyswAphoYAEdcAcvKAAAA
bAEaGQBHXAHLywAAAEQCGhoAR1wBy80AAADsAhobAEdcAcvOAAAApAMaHABHXAHL8AAAAEwEGh0A
R1wBy88AAAAEBRoeAEdcAcvQAAAArAUaHwBHXAHL0QAAAGQGGiAAR1wBy+4AAAAMBxohAEdcAcvM
AAAAxAcaIgBHXAHL7wAAAHwIGiMAHRooAB0aLAAdGjAAHRo0AB0aOAAdGjwAHRpAAB0aRAAdGkgA
HRpMAB0aUAAdGlQAHRpYAB0CAAAAEQBIAAAAAAD//wEAAwADAAAAEgBwARQA4AElAPgBKQAQAi0A
NAIxAFgCNQB0AjkAkAI9AKgCQQDAAkUA2AJJAPQCTQA8AFEAWABVAAoAAAA=
</rawitemdata></item>
```

**Comment by feather812002** [comment]
I still have a idea how to shoot our trouble on the share filed and share action about the raw data encode from binary format.
The idea is   : 
1. first we generate a normal dxl with these design element ,get a nsf database
2. we export the nsf database to binary dxl for these design element .
3. we copy the correct raw encode data from this binary dxl file into our Moonshine-IDE dxl file.

I'm confident enough that the plan will work as expect.

**Comment by feather812002** [comment]
@JoelProminic 
I almost done the save the global option into dxl, 
but I have some question about this :

I see these Lotus script always around by :

```
Sub {Function name}
End Sub
```

So , do we should force add these to script ?
because in the code editor ,people can easy remove it and input some wrong format.

**Comment by JoelProminic** [comment]
> @JoelProminic I almost done the save the global option into dxl, but I have some question about this :
> 
> I see these Lotus script always around by :
> 
> ```
> Sub {Function name}
> End Sub
> ```
> 
> So , do we should force add these to script ? because in the code editor ,people can easy remove it and input some wrong format.

I have a [writeup for the default values](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1009#issuecomment-1793057005).  Note that the `End Sub` will need to be added for each of the `Sub` cases

I'd like to see these defaults appear in the editor.  For now, we don't need to do validation on the values after updates, but that is something we can revisit later.

**Comment by JoelProminic** [comment]
> I still have a idea how to shoot our trouble on the share filed and share action about the raw data encode from binary format. The idea is :
> 
>     1. first we generate a normal dxl with these design element ,get a nsf database
> 
>     2. we export the nsf database to binary dxl for these design element .
> 
>     3. we copy the correct raw encode data from this binary dxl file into our Moonshine-IDE dxl file.
> 
> 
> I'm confident enough that the plan will work as expect. The difficult is how to accurately identify raw encoded data of design elements from source dxl file.

This sounds like a potential solution for #1014 and #1184, but I'm not sure if it would be a feasible solution for this case.  You could try creating a temporary SharedAction or Agent with the form code and copying it, but I'm not sure if it would work.

That said, I'm wondering if we would be better off switching to the normal DXL format for the forms, since this is at least the third time we have been blocked by the note format.  Some concerns here:
- I remember there was some reason why we had to switch to the note format, but I don't remember the details.  If we can figure out the original problem, we'll need to decide if it is easier to fix that problem or to fix the problems we are currently seeing.
- Switching to the normal format could be time consuming, and we would need to retest the various form features.

**Comment by JoelProminic** [comment]
Looking at the example normal DXL format [here](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1009#issuecomment-1824335030), I see that all the LotusScript is assigned to a single `$$FormScript` element.  This is then broken up into different "Objects" when it is displayed in the form.

This reminded me that we had added `Grails4NotesPropertiesDONOTMODIFY` and `Grails4NotesCommentDONOTMODIFY` methods to the form-level LotusScript for Structure4Notes.  I see this shows up as a new object:

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/2a7ac2f8-fdcb-41e1-93b4-cddfbec6f4cf)

Revisiting this in Designer, I found this behavior:
1. Select the Terminate object
2. On a new line, define a new method.  
    - as soon as I hit Enter, a new object is created with the new method name.  "End Sub" is added automatically.
    - The Terminate object still exists, with an empty method
3. Select the Terminate object again
4. Change the method name to Terminate2 and hit enter
    - A new Terminate2 object is created
    - The Terminate object remains, with an empty method
5. Delete the contents of the Terminate2 object and hit Enter
    - The Terminate2 method is removed.

**Comment by feather812002** [comment]
@JoelProminic  So, do we need similar this action let the Object tree also can add a new object or remove a new object option in the tree node ?

**Comment by feather812002** [comment]
And other question about the form option , can  I remove source option, for example : "QuerySave" and "PostSave" from the tree in the Designer ?

**Comment by JoelProminic** [comment]
Checking the normal format for the DXL, I see the JavaScript and Formula code here:

```
  <code event="windowtitle">
    <formula>"Test Title"</formula>
  </code>
  <code event="webquerysave">
    <formula>@Command([ToolsRunMacro]; "SaveAgent")</formula>
  </code>
  <code event="webqueryopen">
    <formula>@Command([ToolsRunMacro]; "OpenAgent")</formula>
  </code>
  <code event="htmlbody">
    <formula>"Example Body Attributes"</formula>
  </code>
  <code event="htmlhead">
    <formula>REM {Example Head Content};
"Test"</formula>
  <code event="targetframe">
    <formula>"Test Frame"</formula>
  </code>
  <code event="jsheader" for="web">
    <javascript>"TestJSHeader"</javascript>
  </code>
  <code event="onHelp" for="web">
    <javascript>"Test onHelp"</javascript>
  </code>
  <code event="onLoad" for="web">
    <javascript>"Test onLoad"</javascript>
  </code>
  <code event="onUnload" for="web">
    <javascript>"Test onUnload"</javascript>
  </code>
  <code event="onClick" for="web">
    <javascript>"Test onClick"</javascript>
  </code>
  <code event="onDblClick" for="web">
    <javascript>"Test onDblClick"</javascript>
  </code>
  <code event="onKeyDown" for="web">
    <javascript>"Test onKeyDown"</javascript>
  </code>
  <code event="onKeyPress" for="web">
    <javascript>"Test onKeyPress"</javascript>
  </code>
  <code event="onKeyUp" for="web">
    <javascript>"Test onKeyUp"</javascript>
  </code>
  <code event="onMouseDown" for="web">
    <javascript>"Test onMouseDown"</javascript>
  </code>
  <code event="onMouseMove" for="web">
    <javascript>"Test onMouseMove"</javascript>
  </code>
  <code event="onMouseOut" for="web">
    <javascript>"Test onMouseOut"</javascript>
  </code>
  <code event="onMouseOver" for="web">
    <javascript>"Test onMouseOver"</javascript>
  </code>
  <code event="onMouseUp" for="web">
    <javascript>"Test onMouseUp"</javascript>
  </code>
  <code event="onReset" for="web">
    <javascript>"Test onReset"</javascript>
  </code>
  <code event="onSubmit" for="web">
    <javascript>"Test onSubmit"</javascript>
  </code>
```

With `note` format, I see the formula entries like this, except for targetFrame:

```
<item name='$WindowTitle' sign='true'><formula>"Test Title"</formula></item>
<item name='$HTMLHeadTag' sign='true'><formula>REM {Example Head Content};
"Test"</formula></item>
<item name='$HTMLBodyTag' sign='true'><formula>"Example Body Attributes"</formula></item>
<item name='$WEBQuerySave' sign='true'><formula>@Command([ToolsRunMacro]; "SaveAgent")</formula></item>
<item name='$WEBQueryOpen' sign='true'><formula>@Command([ToolsRunMacro]; "OpenAgent")</formula></item>
```


However, I don't see the JavaScript code directly.   After some experimentation with decoding the base 64, I can see the JavaScript string literals in the `$HTMLCode` item.  I'm still not clear on how the encoding works otherwise, though.

```
<item name='$HTMLCode' sign='true'>
<rawitemdata type='1'>
...
```

On a side note, I used similar methods to identify how the LotusScript fields are linked:

Plain Text (included in note format) | Binary | Notes
-- | -- | --
`$Script` | `$$Script_O` | Headers + Initialize and Terminate
`$$FormScript` | `$$$FormScript_O` | Other Subs

**Comment by JoelProminic** [comment]
I confirmed that the `1009` error from yesterday is resolved, and I can view and edit the objects properties.

I'm still seeing this error when saving a form.  It didn't trigger every time, so I'm trying narrow down the conditions:

    : TypeError: Error #1090
    : 	at components.domino::DominoGlobalsObjects/toCode()
    : 	at view.domino.surfaceComponents.components::DominoGlobalsObjects/toCode()
    : 	at actionScripts.plugin.dominoInterface::DominoObjectsPlugin/handleDominoObjectsSave()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at actionScripts.plugins.ui.editor::VisualEditorViewer/save()
    : 	at actionScripts.controllers::SaveFileCommand/execute()
    : 	at actionScripts.locator::IDEController/execCommand()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at actionScripts.utils::KeyboardShortcutManager/dispatch()
    : 	at actionScripts.utils::KeyboardShortcutManager/stopEvent()
    : 	at actionScripts.ui.menu::MenuPlugin/redispatch()
    : 	at Function/<anonymous>()
    : 	at actionScripts.impls::INativeMenuItemBridgeImp/_trigger()
    : 	at flash.display::NativeMenuItem/performKeyEquivalent()
    : 	at flash.display::NativeMenu/performKeyEquivalent()
    : 	at flash.display::NativeMenuItem/performKeyEquivalent()
    : 	at flash.display::NativeMenu/performKeyEquivalent()
    : 	at flash.desktop::NativeApplication/_onKeyDownCapture()
    :
    : Click here to Report a Bug


I also saw this behavior:
1. Create a fresh project and form
2. Add a label and field to the form
3. Add comments to the "(Globals)" objects, with an example of special XML characters
4. Reopen the form.  The Object values are preserved
5. Create a new form. BUG:  It shows the same object values
6. Save the new form.  I see that the options for the first form are added for the new form
7. Close all forms and restart Moonshine
8. Create a new form.  The options are displayed properly
9. Open the default form.  The saved options are displayed
10. Switch back to the new form. BUG:  it still shows the modified options

**Comment by JoelProminic** [comment]
I confirmed the default types match [what I expect](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1009#issuecomment-1793055068) for a new form:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/fdbaf130-08c7-4e05-bd32-6f21264ef0b9)


Expected:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/450ba7aa-7361-40c7-b7d3-e4e0dd0233ca)

Note that the objects with default values did not show as populated in Designer - these values were apparently populated when opening the object, and saved only if the value was changed.  This is not a blocking bug, unless the default methods will cause errors in the database (for example, WebQuerySave and WebQueryOpen).

**Comment by JoelProminic** [comment]
I checked the objects from [this table](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1009#issuecomment-1793097714), and it looks like they all have the correct Environment and Language options.

However, I see that the unmentioned JavaScript entries (onClick through onRest) have both Client and Web environments:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/b3451dc1-a6d1-4b51-8550-673516a84c71)

Checking Designer, these should have "Web" as the only option, and the combobox should be disabled:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/e1194459-a044-491e-bd23-3b0b92be14b0)

**Comment by JoelProminic** [comment]
(NOTE:  Old message that was not submitted)  I confirmed that the JavaScript properties with only the "Web" environment have been fixed.  I also confirmed that the other JavaScript cases were properly updated to show "Web" as the default environment.

**Comment by JoelProminic** [comment]
I set values for all the options and build the database.

When I opened the form in the Notes client, I saw 4-5 popups like this before the form opened, and 1 when I closed the form.

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/cb98cf0f-5996-48ff-84fe-0942008fb100)

I copied the database to a Windows workstation and tried to open it with Domino Designer.  I got othis error:

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/b06a5f2f-6ad3-4a93-aab6-8b80c27ca223)

I uploaded this to blazeds (A55D78/MoonshineTests/temp/TestDominoVisualEditor_1222_2.nsf) and then opened it in Designer.  The database opened, but I saw this error on the console;

    Description	Resource	Path	Location	Type
    Global Module : Unexpected: SUB; Expected: Expression; End-of-statement	TestDominoVisualEditor_1222.form	/TestDominoVisualEditor_1222_2.nsf/Forms	line 12	LotusScript Problem


I was able to open the form, and I checked the options.  I found that none of the Formula or JavaScript options were populated, but I think this is expected for now.  I got errors like this when I checked the LotusScript options:

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/c0006a65-98ea-4623-b85a-1d95f00c2dcd)
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/ce6127d2-4e9f-4d13-b41a-f6ca94e8edbd)

Individual problems:
- "(Options)" and "(Declarations)" were not preserved at both the Global and Form level
- "(Globals) > Terminate" was duplicated.  I noticed that there was a commented `' Sub Terminate` line - I don't think I added this.
- I see an extra "BindEvents" option
- Many of the LotusScript options were duplicated

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/92e273de-e2ef-49e6-8b38-6caf2dc69f94)


Otherwise, most of the options seemed to be populated properly, but the errors made it difficult to do consisten testing.

Here is my test project, including the compiled database.

[TestDominoVisualEditor_1222.zip](https://github.com/Moonshine-IDE/Moonshine-IDE/files/13757201/TestDominoVisualEditor_1222.zip)

**Comment by feather812002** [comment]
@JoelProminic 
I test this in today , it look like it cause the global options convert to dxl have some misunderstand , 
As we know , our correct LotusScript format is :

```
Sub Initialize
	' Test Globals.initialize
End Sub
```

So , I add some logic in here, if the code no found the "Sub Initialize" in the text it will auto add it .
But it look like some time it will broken this , 
For example, if you input : 

```
Sub Initial
	' Test Globals.initialize
End Sub
```

The logic will auto add the "Sub Initialize"
```
Sub Initialize
'Sub Initial
	' Test Globals.initialize
End Sub
```

So, this maybe cause some trouble , I remove this logic at now , and so you must should make sure you input the correct format of LotusScipt, otherwise it must broken the NSF database build again.

**Comment by JoelProminic** [comment]
I tested this again today, and I got a different error in the Notes client.  The error makes me think it is a problem with `(Globals)` or `(Declarations)`:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/f176215e-7280-43db-b379-931beb64fbf4)

I uploaded the database to blazeds, A55D78/MoonshineTests/temp/TestDominoVisualEditor_1228_2.nsf.

Checking in Designer, I get this error when clicking on `(Globals)` or `(Declarations)`, which prevents me from checking the value:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/53b8c0bc-377f-4f0c-982b-4c2cbe1104bb)

I see this error reported at the bottom:

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/b85b65bb-f255-47f5-aec2-8444125a12c1)

**Comment by feather812002** [comment]
@JoelProminic 
Today , I done the install all Designer on the windows VM , and starting work your test project,
I just copy your test nsf database to the windows VM, rather than copy it to blazeds.testprominic.com.
I can success open it no any other error, but it show a red cross symbol , as follow:

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/c6e93226-309d-4158-a2a6-88c3e36403ad)

And I can see the Lotus Script function fill into correct, 
but it pop "LSE Failuer" error message as your same on above, beside this, I no see any other error:

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/ba7c93ba-c21e-471f-80d8-ac31a3cc903e)


I will investigate why it get the "LSE Failuer" error, thanks.

**Comment by feather812002** [comment]
@JoelProminic 
Today, I test some times with Designer and Moonshine-IDE, final I found we have a format error in the Global options setting, 
From your test project ,I see this code : 



```
Declarations Public
' Test.Globals.Declarations
```

But in Designer it alert this is wrong format and can't be save.
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/fd594647-4609-40a2-95c3-1d32bde49c4c)

So , we should remove this code,and I still not understand the Designer how check the format , maybe it embed a LotusScript compiler inside it , I not sure we can exact it from Designer and embed it into the Moonshine-IDE.
But I think even we can do this , it maybe bring some copyright issues.

So, what do you think about this ?

If we not do the format check , in the Moonshine-IDE editor user can input any code , and it still maybe compile success, 
but when open it with Designer or move it to Domino server it must broken ,and can't be execute.

**Comment by feather812002** [comment]
@JoelProminic 

About the compile LotusScript, I think ODP already to do similar work about this :  

> These libraries are brought in with their source stored as [normal text items](https://github.com/OpenNTF/org.openntf.nsfodp/blob/e595156c49cda91f74e344869ee70c616f2d0b39/eclipse/plugins/org.openntf.nsfodp.compiler/src/org/openntf/nsfodp/compiler/ODPCompiler.java#L901) named $ScriptLib, but then need to be compiled from there. There's no mechanism for compiling LotusScript in the normal Java API, and IBM's NAPI doesn't have a binding to the [NSFNoteLSCompileExt](http://www-12.lotus.com/ldd/doc/domino_notes/8.5.3/api853ref.nsf/f961929dc6dc14dc852561bf0067d1c4/fa422b61d9afd014482570bb0056c05d?OpenDocument) function involved, so I had to [dip into Java C API bindings](https://github.com/OpenNTF/org.openntf.nsfodp/blob/e595156c49cda91f74e344869ee70c616f2d0b39/eclipse/plugins/org.openntf.nsfodp.compiler/src/org/openntf/nsfodp/compiler/ODPCompiler.java#L929), initially via Karsten Lehmann's excellent [Domino JNA](https://github.com/klehmann/domino-jna) and then switched over to [Darwino's NAPI implementation](https://github.com/darwino/domino-napi).

[More Details](https://frostillic.us/blog/posts/2019/7/2/deab5e75ce7c5bdf8525842b004f6c00)

So , this is my design about this , compile Lotus Script and it's  library in the NSF is possible .

We already have a REST interface from NSFConvertProtal VM, and it can build nsf database with ODP.

So, in short , we just change some code inside of the ODP,let it support REST call and return the result to REST,
Then we can got the compile result and send it to Moonshine-IDE by REST .

**Comment by feather812002** [comment]
I to so some research about the NsfODP project how to compile the LotusScript inside.
Now, I almost understand NSFODP at least support follow type element to compile with Lotus Script.
- LotusScript library [LotusScript library](https://github.com/OpenNTF/org.openntf.nsfodp/blob/836fd0ff389e452fd66c7ac46af6f507d91cf359/nsfodp/bundles/org.openntf.nsfodp.compiler/src/org/openntf/nsfodp/compiler/ODPCompiler.java#L824)
-  [Other element that contain Lotus Script also can be compile but it need some setting](https://github.com/OpenNTF/org.openntf.nsfodp/blob/836fd0ff389e452fd66c7ac46af6f507d91cf359/nsfodp/bundles/org.openntf.nsfodp.compiler/src/org/openntf/nsfodp/compiler/ODPCompiler.java#L650)
-  [Database that contain the library](https://github.com/OpenNTF/org.openntf.nsfodp/blob/836fd0ff389e452fd66c7ac46af6f507d91cf359/nsfodp/bundles/org.openntf.nsfodp.compiler/src/org/openntf/nsfodp/compiler/ODPCompiler.java#L853)

**Comment by feather812002** [comment]
JavaScript check in Designer.
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/6170de09-3883-46ce-90bf-e2ca9c54637a)
As you see ,if you input a wrong format of JavaScript , it will reject your save action ,
For formula ,it support a list formula prompt 
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/428e2f67-e3a8-4f20-93ac-6353ec093966)
And if you input a wrong formula , it also report a bug and not allow you save:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/e0093ac4-2328-4516-a4d3-7fcdb40cb803)


So for next , we should consider if we want support these language verify and check.

**Comment by feather812002** [comment]
About JavaScript int the dxl, I think already found this format as follow:

> <item name='$HTMLCode' sign='true'>
> <rawitemdata type='1'>
> +f8gAAAAAAABAAQAOQAAAAAAAAAAAAAAAAAAAAAAAADc/0wA+f86ACBOAAAAAAAAAABmdW5jdGlv
> biBhbGVydG1lc3NnKCl7DSAgICBjb25zb2xlLmxvZygiaGVsbG8sd29ybGQiKTsNfQAA+f8gAAAA
> AAANAAQADQAAAAAAAAAAAAAAAAAAAAAAAADc/yAA+f8OACBOAAAAAAAAAAB2YXIgdGVzdD0iIjsA
> bV//EAACAA8ABwAAAAAAAABf/xAAAgANAAcAAAAAAAAA
> </rawitemdata></item>

It look like all JS code will encode into the "$HTMLCode" item.

The trouble is  , it insert other some code into this segment,and encode it  ,
I try to decode it with base64, the JS code is decode correct , but other some code is code segment.

source JS code:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/776428b3-47e3-4f50-843e-79aade451714)


After decode from this itme:
 �������9�������������������L�:� N��������function alertmessg(){
    console.log("hello,world");
}�� �����
��
������������������� �� N��������var test="";�m_����������_��
��������

**Comment by JoelProminic** [comment]
I tried a new build today, and the focus in/out errors were fixed (see #1248).

However, now the Objects tab is not showing up, and I get this error when saving.

    : Error #1009
    : TypeError: Error #1009
    : 	at actionScripts.plugin.dominoInterface::DominoObjectsPlugin/handleDominoObjectsSave()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at actionScripts.plugins.ui.editor::VisualEditorViewer/save()
    : 	at actionScripts.controllers::SaveFileCommand/execute()
    : 	at actionScripts.locator::IDEController/execCommand()
    : 	at flash.events::EventDispatcher/dispatchEvent()
    : 	at actionScripts.utils::KeyboardShortcutManager/dispatch()
    : 	at actionScripts.utils::KeyboardShortcutManager/stopEvent()
    : 	at actionScripts.ui.menu::MenuPlugin/redispatch()
    : 	at Function/<anonymous>()
    : 	at actionScripts.impls::INativeMenuItemBridgeImp/_trigger()
    : 	at flash.display::NativeMenuItem/performKeyEquivalent()
    : 	at flash.display::NativeMenu/performKeyEquivalent()
    : 	at flash.display::NativeMenuItem/performKeyEquivalent()
    : 	at flash.display::NativeMenu/performKeyEquivalent()
    : 	at flash.desktop::NativeApplication/_onKeyDownCapture()

**Comment by feather812002** [comment]
I can't repeat this error on my side ,
Maybe I work on the old version Text editor ,
Let us discuss in tonight and try to repeat it .

**Comment by feather812002** [comment]
For normal dxl format about javascript 

```
<?xml version='1.0' encoding='utf-8'?>
<form name='newForm' xmlns='http://www.lotus.com/dxl' version='12.0' maintenanceversion='1.0'
 replicaid='C2258AB7003FBBBE' publicaccess='false' designerversion='8.5.3'
 renderpassthrough='true'>
<noteinfo noteid='1b6' unid='9B81A5A6D10EA620C2258ABB0043FA58' sequence='2'>
<created><datetime dst='true'>20240206T152230,00+03</datetime></created>
<modified><datetime dst='true'>20240206T152248,77+03</datetime></modified>
<revised><datetime dst='true'>20240206T152248,76+03</datetime></revised>
<lastaccessed><datetime dst='true'>20240206T152248,76+03</datetime></lastaccessed>
<addedtofile><datetime dst='true'>20240206T152230,13+03</datetime></addedtofile></noteinfo>
<updatedby><name>CN=Demo Admin/O=DEMO</name></updatedby>
<wassignedby><name>CN=Demo Admin/O=DEMO</name></wassignedby><code event='onClick'
 for='web'><javascript>test</javascript></code>
<body><richtext>
<pardef id='1'/>
<par def='1'/></richtext></body>
<item name='$$ScriptName' summary='false' sign='true'><text>newForm</text></item></form>
```

**Comment by feather812002** [comment]
so in short final we will call JNA API convert the 
```
<code event='onClick'
 for='web'><javascript>test</javascript></code>
```
To

```
+f8gAAAAAAABAAQABQAAAAAAAAAAAAAAAAAAAAAAAADc/xgA+f8GACBOAAAAAAAAAAB0ZXN0ACJf
/xAAAgAPAAcAAAAAAAAA
```

**Comment by feather812002** [comment]
For common JavaScript 


```
<?xml version='1.0' encoding='utf-8'?>
<form name='FormProperty' xmlns='http://www.lotus.com/dxl' version='12.0'
 maintenanceversion='1.0' replicaid='C2258AB7003FBBBE' publicaccess='false'
 designerversion='8.5.3' renderpassthrough='true'>
<noteinfo noteid='13a' unid='465B7A74ECA0CA52C2258AB7003FD34C' sequence='10'>
<created><datetime dst='true'>20240202T143708,60+03</datetime></created>
<modified><datetime dst='true'>20240206T171848,52+03</datetime></modified>
<revised><datetime dst='true'>20240206T171848,51+03</datetime></revised>
<lastaccessed><datetime dst='true'>20240206T171848,51+03</datetime></lastaccessed>
<addedtofile><datetime dst='true'>20240202T143710,08+03</datetime></addedtofile></noteinfo>
<updatedby><name>CN=Demo Admin/O=DEMO</name></updatedby>
<wassignedby><name>CN=Demo Admin/O=DEMO</name></wassignedby><code event='onClick'
 for='web'><javascript>test</javascript></code><code event='onLoad' for='web'><javascript
>commonjs</javascript></code><code event='onDblClick' for='web'><javascript
>1111</javascript></code><code event='onHelp' for='web'><javascript>common</javascript></code><code
 event='onHelp' for='client'><javascript>common</javascript></code><code event='onLoad'
 for='client'><javascript>commonjs</javascript></code>
<body><richtext>
<pardef id='1'/>
<par def='1'/></richtext></body>
<item name='$$ScriptName' summary='false' sign='true'><text>FormProperty</text></item></form>
```

**Comment by feather812002** [comment]
```
<code event='onClick'
 for='web'><javascript>onclick</javascript></code><code event='onLoad' for='web'><javascript
>commonjs</javascript></code><code event='onDblClick' for='web'><javascript
>onDblClick</javascript></code><code event='onHelp' for='web'><javascript
>onhelp</javascript></code><code event='jsheader' for='web'><javascript>JSheader</javascript></code><code
 event='onUnload' for='web'><javascript>onunload</javascript></code><code
 event='onKeyDown' for='web'><javascript>onKeyDown</javascript></code><code
 event='onKeyPress' for='web'><javascript>onKeyPress</javascript></code><code
 event='onKeyUp' for='web'><javascript>onKeyUp</javascript></code><code event='onMouseDown'
 for='web'><javascript>onMouseDown</javascript></code><code event='onMouseMove'
 for='web'><javascript>onMouseMove</javascript></code><code event='onMouseOut'
 for='web'><javascript>onMouseOut</javascript></code><code event='onMouseOver'
 for='web'><javascript>onMouseOver</javascript></code><code event='onMouseUp'
 for='web'><javascript>onMouseUp</javascript></code><code event='onReset'
 for='web'><javascript>onRest</javascript></code><code event='onSubmit' for='web'><javascript
>onSubmit</javascript></code><code event='onHelp' for='client'><javascript
>onhelp</javascript></code><code event='onLoad' for='client'><javascript>onload</javascript></code><code
 event='jsheader' for='client'><javascript>JSheader</javascript></code><code
 event='Querymodechange' for='client'><javascript>QueryModeChange</javascript></code><code
 event='Postmodechange' for='client'><javascript>PostModeChange</javascript></code><code
 event='Queryrecalc' for='client'><javascript>Queryrecalc</javascript></code><code
 event='Postrecalc' for='client'><javascript>Postrecalc</javascript></code><code
 event='Postsave' for='client'><javascript>postsave</javascript></code><code
 event='Querysend' for='client'><javascript>querysend</javascript></code><code
 event='Postsend' for='client'><javascript>postsend</javascript></code><code
 for='client'><javascript>onSize</javascript></code>
```

**Comment by feather812002** [comment]
@JoelProminic 
I want reproduce error on my side base on your condition.
- I download the install file from https://bamboo.prominic.net/browse/MS-BMOB-936/artifact/shared/MacOS-Installers/
- Install it on my local work station:

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/7b3da907-64f0-4262-b79a-41936ee4f3ee)
and create a new "Domino Visual NSF project" , then create a new form .
- Change the form property 
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/46a074d3-4250-467d-be34-3bc53fa9eca0)
- No any error throw and save success

So, I can't reproduce your error even I test on the build from bamboo, 
I suggestion you test on other workstation in your local.

**Comment by JoelProminic** [comment]
After some more discussion, we were able to reproduce the problem like this
1. Restart Moonshine
2. Open a form
3. Save.  BUG:  the error triggers
4. Save.  BUG:  the error triggers again
5. Update an option 
6. Save.  No error
7. Reopen the form and save.  No error
8. Create a new form and save. No error
9. Restart Moonshine
10. Open a form
11. . Save.  BUG:  the error triggers again

So, this error stops triggering once an object updated once, and this seems to clear the error until Moonshine is next started.   I want to fix this error - I would expect that most users (especially in the short term) will not touch the Objects interface.

**Comment by feather812002** [comment]
@JoelProminic 
I confirm this error can't trigger on the last build 940
Try to test it in today, I think this error will be fixed already

**Comment by JoelProminic** [comment]
Testing with the latest build (940), I confirmed that the save error is fixed.  However, I am getting LotusScript errors when opening the generated forms, even if I don't change the Objects.
1. Create a new project.  The form opens automatically
2. Add a Label and Field.  Use "TestID" for both.
3. Save the form
4. Project > Build on Vagrant
5. Browse to nsfs/nsf-moonshine/target and open the generated database
6. In the `All By UNID/CRUD/<form>` view, click New.
7. BUG:  Error triggers:  "Unexpected: Public; Expected Expression; End-of-statement; ,; BYVAL

We need to make sure that the users can create forms without touching the Objects interface if we want to merge this for the 3.3.4 release.

**Comment by JoelProminic** [comment]
I also created a form where I set all of the objects.  This triggered the same errors as above when opening and closing the form 

I uploaded the database here:
- blazeds.testprominic.com/PNITEST
- A55D78/MoonshineTests/temp/TestDominoVisualEditor_0209.nsf
- Form: TestDominoVisualEditor_0209

Checking the form in Designer, I found that the LotusScript objects were preserved properly, except for the form Initialize and Terminate.  The Formula and JavaScript objects were not preserved, but this was expected with the current limitations.

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/24532731/cf6cd55d-d543-4dd3-8c7c-77d73b915e44)

**Comment by feather812002** [comment]
I think I already address this error , the default value for (Declarations) is wrong, it should be empty default value.
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/16f0534b-9c9d-4b74-abc8-5137e34171ac)

I will fix it and commit in soon.

**Comment by feather812002** [comment]
OKay, I think I found a new issue(maybe not a issue)
If user no input anything in the form with Moonshine-IDE, the form not work in the Designer , it will report "Not a form"!
I check the dxl , it look like no error , 
after I add any other designer element (one field or one label )it work fine and no any error.

**Comment by feather812002** [comment]
It look like whatever the Form or Page all both should contain at least `<par>` element, otherwise , it will report error in Designer or Notes when you try to open it.
 I already fix this and commit it

**Comment by feather812002** [comment]
Today I work staring with save the formula into dxl and build it back to NSF database.
Now, I post the a example dxl about the formula format in here:


```
<?xml version='1.0' encoding='utf-8'?>
<note class='form' xmlns='http://www.lotus.com/dxl' version='12.0' maintenanceversion='1.0'
 replicaid='C2258AB7003FBBBE'>
<noteinfo noteid='13a' unid='465B7A74ECA0CA52C2258AB7003FD34C' sequence='23'>
<created><datetime dst='true'>20240202T143708,60+03</datetime></created>
<modified><datetime dst='true'>20240214T163639,43+03</datetime></modified>
<revised><datetime dst='true'>20240214T163639,42+03</datetime></revised>
<lastaccessed><datetime dst='true'>20240214T163639,42+03</datetime></lastaccessed>
<addedtofile><datetime dst='true'>20240202T143710,08+03</datetime></addedtofile></noteinfo>
<updatedby><name>CN=Demo Admin/O=DEMO</name></updatedby>
<wassignedby><name>CN=Demo Admin/O=DEMO</name></wassignedby>
<item name='$Comment' sign='true'><text/></item>
<item name='$HTMLCode' sign='true'>
<rawitemdata type='1'>

</rawitemdata></item>
<item name='$TITLE'><text>FormProperty</text></item>
<item name='$Info' sign='true'>
<rawitemdata type='1'>
hhgBAIAAAAAAgAAAAQABAP///wAQAAAA
</rawitemdata></item>
<item name='$WindowTitle' sign='true'><formula>"WindowsTitle"</formula></item>
<item name='$HTMLHeadTag' sign='true'><formula>"HTMLHead"</formula></item>
<item name='$HTMLBodyTag' sign='true'><formula>"HTMLBODY"</formula></item>
<item name='$WEBQuerySave' sign='true'><formula>@Command([ToolsRunMacro]; "openSave")</formula></item>
<item name='$WEBQueryOpen' sign='true'><formula>@Command([ToolsRunMacro]; "openWeb")</formula></item>
<item name='$Flags'><text/></item>
<item name='$DesignerVersion'><text>8.5.3</text></item>
<item name='$Script' summary='false' sign='true'><text>'++LotusScript Development Environment:2:5:(Options):0:74
Option Public


'++LotusScript Development Environment:2:5:(Forward):0:1

'++LotusScript Development Environment:2:5:(Declarations):0:2

<break/></text></item>
<item name='$$Script_O' sign='true'>
<rawitemdata type='14'>
AQAAAkxTT0IDABQIZW4AAAQAAAADAAAAAAAEAIgAAAAAAAAAAAAAAAAAAAAkAAAAGAA4AP////9U
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAEABQBMb3R1cyBQcm9kdWN0AAAAVjAuMAAAAAAAAAAAAAAAAAAA
AAAAAAAAAwAAAAAAAAAEAAAAiAAAAAAAAAAAAAAAAQAAAAAAAAAAAAAAAAAAAAEAAAAAAAAAAAAA
AAMAAABXMzI/JQMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA
AAAAAwCMAAAAAAD//wgAKgA3AEYAOAAyADIAMgA4AAAAAABsAAMATgBFAFcAAACAAAYARABFAEwA
RQBUAEUAAAAAAP//CgBJAE4ASQBUAEkAQQBMAEkAWgBFAAAAAAD//wkAVABFAFIATQBJAE4AQQBU
AEUAAAD//wYATwBCAEoARQBDAFQAAAAAAP//AAAAAAAABQAIAAAAAAAp+tuUBAAHAAAAAAAdAAAC
AAAAEQAQAAAAAAD//wEAAwAFAAAABwAKAAAA
</rawitemdata></item>
<item name='$$ScriptName' summary='false' sign='true'><text>FormProperty</text></item>
<item name='$$FormAction7' sign='true'><formula>"PostSave"</formula></item>
<item name='$Body' sign='true'><richtext>
<pardef id='1'/>
<par def='1'/></richtext></item></note>
```

**Comment by feather812002** [comment]
I done a basic version for save formula to dxl
it look like follow:

![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/a87c1cc8-4ebb-466c-a718-9a80af0ca9f4)

After build NSF database, it show correct with Designer:
![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/acdf4796-2fb6-4d57-a042-9b535cfc7538)


I also test other 3-4 formula property about the formula language work with Objects, it look like all work fine.
I not test all property at now.

**Comment by feather812002** [comment]
- No Lotus Script Compile VM running
   - Fresh project without save  ,build to NSF and open with Notes client
      Done and No error , the View can be open correct and the form can be open correct with project name form
      
      ![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/a40bfae2-dac2-4b83-9923-d8d7f86e0899)

   - Fresh project save with no element ,build to NSF and open with Notes client
      Done and No error same on above test result
   - Fresh project create a new form and no element in the form ,add JS function 
     - build to NSF database and open with Notes client 
       ![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/0cf1406c-44b5-44b8-992b-8970c50d74ca)
     - open with Designer check the JS if save correct
       It can be open Designer correct , but the JS is Empty(because no Lotus Script Compile VM running so ,it can't be convert to raw data , so it is empty )
   - Fresh project create a new form and no element in the form ,add Formula into Object options. 
     - build to NSF database and open with Notes client 
        It can be open correct with Notes client , and the view&form open correct with correct name.
     - open with Designer check the JS if save correct
       ![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/8d1bd1b3-8bfc-4896-b2e9-1061c0b65975)
       As you see, the formula save correct and can be review with Designer.
   - Fresh project create a new form and no element in the form ,add Lotus Script into Object options. 
      - build to NSF database and open with Notes client 
         ![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/979abb4c-9de3-4b20-b0a4-04895348a38c)

      - open with Designer check the LotusScript if save correct 
         ![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/a920d6b9-7931-4958-a87b-b508578f9287)

         ![image](https://github.com/Moonshine-IDE/Moonshine-IDE/assets/8543391/76083803-3204-4eee-bf5c-9e1d2c972021)

---------------------------------------------------

So in short , for current , LotusScript & Formula & JavaScript will all persistence in the Moonshine-IDE side
When you input any thing about these , they will be save into the intermediary xml .
But only running the LotusScript Compile VM, you can save the JavaScript into dxl.

Formula&LotusScript can be directly save into dxl , but if you input a wrong Formula or LS format, it maybe broken the ODPNSF build.

**Comment by feather812002** [comment]
Currently doing the simple tests above, more complex tests, or if you want more edge test cases, we can determine that in tonight's discussion @JoelProminic

**Comment by feather812002** [comment]
I found a way can compile the Formula in today ,it throw follow error information:

`FormulaCompilationError: Formula Error. Unknown @Function, line=0, column=0, offset=0, length=5, formula=@1234`

So,I think it is okay for our current scenes to be used

**Comment by JoelProminic** [comment]
I did some testing [here](https://github.com/Moonshine-IDE/Moonshine-IDE/issues/1234#issuecomment-1986398077), and this should be good for a merge.

**Comment by JoelProminic** [comment]
The remaining issue on #1234 was resolved, so this is ready for a merge.  As usual:
- Merge master in to the branch before merging
- Resolve any conflicts and run a build before pushing the merge.  I don't expect any major problems, since we haven't been doing other development for Domino Visual Editor at the moment.

**Comment by JoelProminic** [comment]
I confirmed that this has been merged.
